### PR TITLE
Refactored code, moved test results caching logic in TestExecutionRecorder

### PR DESCRIPTION
**Pull request by harshjain2**
Moved the logic to cache TestResults in TestExecutionRecorder class. 

As a framework guideline, we should get events in this order :
1. Test Case Start
2. Test Case End
3. Test Case Result. 

This PR handles the case if the order of events doesn't follow the above mentioned order. 

Earlier this logic was was DataCollectors, but it's the responsibility of TestPlatfrom to provide the events in correct order, therefore, moved this logic out of DataCollectors and putting it in TestExecutionRecorder.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs [hunk 1]
```diff
@@ -32,6 +38,9 @@ public TestExecutionRecorder(ITestCaseEventsHandler testCaseEventsHandler, ITest
             this.testRunCache = testRunCache;
             this.testCaseEventsHandler = testCaseEventsHandler;
             this.attachmentSets = new List<AttachmentSet>();
+
+            this.testResultDictionary = new Dictionary<Guid, List<TestResult>>();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs]: Please add a comment explaining why the error correction is required. It will help in future.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs [hunk 2]
```diff
@@ -68,11 +82,37 @@ public void RecordStart(TestCase testCase)
...
+                    }
+                    else
+                    {
+                        // No TestCaseEnd received yet
+                        // We need to wait for testcaseend before flushing
+                        flushResult = false;
+
+                        List<TestResult> testResults;
+                        // Cache results so we can flush later with in proc data
+                        if (this.testResultDictionary.TryGetValue(testCaseId, out testResults))
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs]: Line 104 can be inlined directly with new C# syntax.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs [hunk 3]
```diff
@@ -93,7 +133,41 @@ public void RecordEnd(TestCase testCase, TestOutcome outcome)
...
+                {
+                    isTestCaseEndAlreadySent = this.testCaseEndStatusMap.Contains(testCase.Id);
+                    if (!isTestCaseEndAlreadySent)
+                    {
+                        this.testCaseEndStatusMap.Add(testCase.Id);
+                    }
+
+                    // Do not support multiple - testcasends for a single test case start
+                    // TestCaseEnd must always be preceded by testcasestart for a given test case id
+                    if (!isTestCaseEndAlreadySent)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs]: Can we validate following scenarios please:
* LUT with Theory and DataRow in test frameworks
* Xunit test where ClassInitialize or TestInitialize fails (they send multiple test results in this case too)

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 4]
```diff
@@ -166,6 +161,7 @@ private void TriggerUpdateTestResult(object sender, TestResultEventArgs e)
         {
             // Just set the cached in-proc data if already exists
             this.SetInProcDataCollectionDataInTestResult(e.TestResult);
+            ((InProcDataCollectionSink)this.inProcDataCollectionSink).RemoveDataCollectionDataSetForTestCase(e.TestCaseId);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Why is this change required?

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 5]
```diff
@@ -28,6 +32,10 @@ public class TestRequestHandler : IDisposable, ITestRequestHandler
 
         private Action<Message> OnAckMessageRecieved;
 
+        private ProxyOutOfProcDataCollectionManager outOfProcDataCollectionManager;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Why do we need instances of data collection managers in this class?

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 6]
```diff
@@ -298,5 +306,28 @@ public int LaunchProcessWithDebuggerAttached(TestProcessStartInfo testProcessSta
...
+            ITestCaseEventsHandler testCaseEventsHandler = null;
+
+            if ((XmlRunSettingsUtilities.IsDataCollectionEnabled(runSettings) && DataCollectionTestCaseEventSender.Instance != null) || XmlRunSettingsUtilities.IsInProcDataCollectionEnabled(runSettings))
+            {
+                testCaseEventsHandler = new TestCaseEventsHandler();
+                var testEventsPublisher = testCaseEventsHandler as ITestEventsPublisher;
+
+                if (DataCollectionTestCaseEventSender.Instance != null && XmlRunSettingsUtilities.IsDataCollectionEnabled(runSettings))
+                {
+                    this.outOfProcDataCollectionManager = new ProxyOutOfProcDataCollectionManager(DataCollectionTestCaseEventSender.Instance, testEventsPublisher);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Not sure if this should be a responsibility of TestRequestHandler to register datacollector managers.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs [hunk 7]
```diff
@@ -20,19 +21,29 @@ namespace TestPlatform.CrossPlatEngine.UnitTests.Adapter
...
+        private Mock<ITestCaseEventsHandler> mockTestCaseEventsHandler;
+        private TestExecutionRecorder testRecorder, testRecorderWithTestEventsHandler;
+        private TestCase testCase;
+        private Microsoft.VisualStudio.TestPlatform.ObjectModel.TestResult testResult;
+
         [TestInitialize]
         public void TestInit()
         {
             this.testableTestRunCache = new TestableTestRunCache();
+            this.testRecorder = new TestExecutionRecorder(null,this.testableTestRunCache);
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: nit: fix styling.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs [hunk 8]
```diff
@@ -125,29 +86,104 @@ public void RecordAttachmentsShouldAddToAttachmentSet()
...
+        {
+            this.testRecorderWithTestEventsHandler.RecordStart(this.testCase);
+            this.testRecorderWithTestEventsHandler.RecordEnd(this.testCase, TestOutcome.Passed);
+            this.testRecorderWithTestEventsHandler.RecordResult(this.testResult);
+
+            this.mockTestCaseEventsHandler.Verify(x => x.SendTestResult(testResult), Times.Once);
+        }
+
+        [TestMethod]
+        public void RecordResultShouldFlushIfTestCaseEndWasCalledBefore()
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: We need tests to validate the order of calls: TestCaseStart, TestCaseResult (1 or more), TestCaseEnd

Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: Or probably enhance the test in L181. There isn't a TestCaseStart in there - it doesn't seem to represent real world scenario.

**Review by codito — APPROVED** [review]
Approved with suggestions.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs (hunk 4 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: This is a bug that we introduced as part of previous refactoring. Due to this bug, Cross and Check were not visible in Code Editor when LUT was enabled.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs (hunk 7 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs]: done.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs (hunk 8 above)
Comment (harshjain2) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: done.

Comment (harshjain2) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Adapter/TestExecutionRecorderTests.cs]: Enhanced the test case, haven't added TestCaseStart though, as it is not required for unit testing.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs (hunk 4 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Moved this logic to the right position in `InProcDataCollectionSink`

File: src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs (hunk 3 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs]: General verification done using xunit.integration project. 
Also verified Theory, DataRow and Fixtures.
